sklearn/model_selection/tests/test_split.py

```diff
@@ -2066,3 +2066,86 @@ def test_stratified_splitter_without_y(cv):
     msg = "missing 1 required positional argument: 'y'"
     with pytest.raises(TypeError, match=msg):
         cv.split(X)
+
+
+@pytest.mark.parametrize(
+    "folds,n_groups,group_size",
+    [
+        (333, 333, 1),
+        (2, 41, 5),
+        (3, 5, 16),
+    ],
+)
+def test_group_kfold_is_stable_with_ties(folds, n_groups, group_size):
+    """Verify groups are assigned based on a stable sort, with ties."""
+    groups = np.repeat(np.arange(n_groups), group_size)
+    X = np.arange(n_groups * group_size)
+    gkf = GroupKFold(folds)
+
+    split = [a for a in gkf.split(X, groups=groups)]
+    for i, (_, test) in enumerate(split):
+        expected = X[groups[::-1] % folds == i]
```

Comment:
Can you explain why `groups[::-1] % folds == i` is expected? Just to understand the logic here.

Reply:
In `_iter_test_indices` the stable sort is reversed to get the largest first. When group lengths are all tied that is the same as just reversing the groups. Then the dealing of groups to folds will start with the first group (was last before reverse) and deal in order to each fol, and since argmin breaks ties with the earliest index it deals normally (not snake style)
Also this expected only works because the group labels are just their group index.

Reply:
I see! Can you explain this in the inline comment just for future reference?

Reply:
It would be an equally good choice to have `expected = X[groups % folds == i]`. It would be more consistent with `KFold` in the `group_size=1` case https://github.com/scikit-learn/scikit-learn/pull/29496#discussion_r1684159533 i.e. in the first split the test set start contains the first sample of X. 

Do we want to do this since we are changing the splits of `GroupKFold`, I would say maybe?

Reply:
I think that is definitely a valid expectation and this wojld be a good time to change that if it will be changed. That way the assignments can maintain consistency across versions, barring required changes. It would also require an update to the no ties test I wrote.

If you all think that is a good change, I can update the assignment loop to be reversed to solve this and make a comment as to why.